File: trading_agent/data/indicators.py
"""Small, dependency-light technical indicator helpers.

Deliberately implemented without pandas/numpy so the whole package has
zero hard runtime dependencies beyond the optional `anthropic` client.
"""
from __future__ import annotations


def sma(values: list[float], window: int) -> float | None:
    """Simple moving average of the last `window` values."""
    if len(values) < window:
        return None
    return sum(values[-window:]) / window


def rsi(values: list[float], window: int = 14) -> float | None:
    """Wilder-style relative strength index over the last `window` changes."""
    if len(values) < window + 1:
        return None
    changes = [values[i] - values[i - 1] for i in range(len(values) - window, len(values))]
    gains = [c for c in changes if c > 0]
    losses = [-c for c in changes if c < 0]
    avg_gain = sum(gains) / window
    avg_loss = sum(losses) / window
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def momentum(values: list[float], window: int) -> float | None:
    """Percent change over the last `window` bars."""
    if len(values) < window + 1:
        return None
    base = values[-window - 1]
    if base == 0:
        return None
    return (values[-1] - base) / base


def volatility(values: list[float], window: int = 20) -> float | None:
    """Annualized standard deviation of daily returns over the last
    `window` bars — a simple realized-volatility estimate. This is the
    risk term in the committee's conviction/risk position sizing
    (`committee/performance_tracker.build_scoreboard`): a name with more
    conviction *and* less realized volatility gets more capital per the
    same idea real risk-parity/vol-targeting position sizing uses, not
    equal-weight regardless of how risky each name actually is.
    """
    if len(values) < window + 1:
        return None
    changes = [
        (values[i] - values[i - 1]) / values[i - 1]
        for i in range(len(values) - window, len(values))
        if values[i - 1] != 0
    ]
    if len(changes) < 2:
        return None
    mean = sum(changes) / len(changes)
    variance = sum((c - mean) ** 2 for c in changes) / (len(changes) - 1)
    return variance**0.5 * 252**0.5

File: webapp-vercel/lib/indicators.ts
export function sma(values: number[], window: number): number | null {
  if (values.length < window) return null;
  const slice = values.slice(-window);
  return slice.reduce((a, b) => a + b, 0) / window;
}

export function rsi(values: number[], window = 14): number | null {
  if (values.length < window + 1) return null;
  const changes: number[] = [];
  for (let i = values.length - window; i < values.length; i++) changes.push(values[i] - values[i - 1]);
  const gains = changes.filter((c) => c > 0);
  const losses = changes.filter((c) => c < 0).map((c) => -c);
  const avgGain = gains.reduce((a, b) => a + b, 0) / window;
  const avgLoss = losses.reduce((a, b) => a + b, 0) / window;
  if (avgLoss === 0) return 100.0;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function momentum(values: number[], window: number): number | null {
  if (values.length < window + 1) return null;
  const base = values[values.length - window - 1];
  if (base === 0) return null;
  return (values[values.length - 1] - base) / base;
}

export function sampleStdev(values: number[]): number {
  if (values.length < 2) return 0;
  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  return Math.sqrt(values.reduce((s, v) => s + (v - mean) ** 2, 0) / (values.length - 1));
}
